[DOC] broker tuning (#4827)
strimzi/strimzi-kafka-operator: documentation/modules/managing/con-broker-config-properties.adoc

// This module is included in the following files:
//
// assembly-client-config.adoc

[id='con-broker-config-properties-{context}']
= Kafka broker configuration tuning

[role="_abstract"]
Use configuration properties to optimize the performance of Kafka brokers.
You can use standard Kafka broker configuration options, except for properties managed directly by Strimzi.

== Basic broker configuration
Certain broker configuration options are managed directly by Strimzi:

* `broker.id` is the id of the Kafka broker
* `log.dirs` is the directory for log data
* `zookeeper.connect` specifies the ZooKeeper connection

Other configuration options are managed through the `Kafka` custom resource:

* `listener` configuration to expose the Kafka cluster to external clients
* `authorization` mechanisms to allow or decline actions executed by users

> * `authentication` mechanisms to prove the identity of users requiring access to Kafka

Author: I've joined together and added a bit more detail for broker IDs and log directories.

Acknowledgments defaults to -1, which means that messages are not acknowledged until replicas for the topic are in sync.

[source,yaml]
----
# ...
offsets.topic.num.partitions=50
offsets.topic.replication.factor=3
offsets.commit.required.acks=-1
# ...
----

*Do not reduce these settings in production, or override the commit acknowledgment.*
You can increase the settings in production.
If you are testing in a single broker environment, you might want to reduce the settings as an exception.

== Improving request handling throughput by increasing I/O threads

Network threads handle requests to the Kafka cluster, such as produce and fetch requests from client applications.
Produce requests are placed in a request queue. Responses are placed in a response queue.

The number of network threads should reflect the replication factor and the levels of activity from client producers and consumers interacting with the kafka cluster.
If you are going to have a lot of requests, you can increase the number of threads. To reduce congestion and regulate the request traffic, you can limit the number of requests allowed in the request queue before the network thread is blocked.

> I/O threads pick up requests from the request queue to process them.

Author: @tombentley - to be confirmed

Adding more threads can improve throughput, but the number of CPU cores/and disk bandwidth imposes a practical upper limit.
The number of threads should be at least the number of disks.

[source,yaml]
----
# ...
num.network.threads=3 <1>
queued.max.requests=500 <2>
num.io.threads=8 <3>
num.recovery.threads.per.data.dir=1 <4>
# ...
----
<1>  The number of I/O  threads for a Kafka broker.
<2>  The number of requests allowed in the request queue.
<3> The number of network threads for the Kafka cluster.
<4> The number of threads used for log loading at startup and flushing at shutdown.

NOTE: Kafka broker metrics can help with working out the number of threads required.
For example, metrics for the average time network threads are idle (`kafka.network:type=SocketServer,name=NetworkProcessorAvgIdlePercent`) indicate the percentage of resources used.
If there is 0% idle time, all resources are in use, which means that more threads would be beneficial.

If threads are slow or limited due to the number of disks, you can try increasing the size of the buffers for network requests to improve throughput:

[source,yaml]
----
# ...

> replica.socket.receive.buffer.bytes=65536

Author: @tombentley - to be confirmed

.Inline messaging flow
image::tuning/broker-tuning-messaging-inline.png[Image of inline messaging flow]

Inline messaging has a performance overhead on the consumer side because of the buffering required, particularly when handling a series of large messages in parallel.
The chunks of large messages can become interleaved, so that it  is not always possible to commit when all the chunks of a message have been consumed if the chunks of another large message in the buffer are incomplete.
For this reason, the buffering is usually supported by persisting message chunks or by implementing commit logic.

.Configuration to handle larger messages

If larger messages cannot be avoided, and to avoid blocks at any point of the message flow, you can increase the message limits with broker configuration.

Use `message.max.bytes` to set the maximum record batch size (which can also be set in topic configuration using `max.message.bytes`).

The broker will reject any message that is greater than the limit set with `message.max.bytes`.
The buffer size for the producers (`max.request.size`) and consumers (`message.max.bytes`) must be able to accommodate the larger messages.

> == Increasing the log flush to improve latency

Author: Increasing the frequency. I've changed the title to _Controlling the log flush of message data_ 
We recommend to use default flush settings, but maybe increasing the frequency with these settings if needed.

In a production environment, to avoid accidental topic deletion resulting in data loss you can also disable deletion. You can temporarily enable it, delete topics and then disable it again.
If you wish to delete topics using the `KafkaTopic` resource, make sure that `delete.topic.enable` is enabled.

[source,env]
----
# ...
auto.create.topics.enable=false
delete.topic.enable=true
# ...
----

== Internal topic settings for transactions and commits

If you xref:reliability_guarantees[configure your producers for reliability guarantees] using transactions for exactly-once writes to multiple partitions, the state of the transactions is stored in the internal `__transaction_state` topic.
By default, the brokers are configured with a replication factor of 3 and a minimum of 2 in-sync replicas for this topic, which means that a minimum of three brokers are required in your Kafka cluster.

[source,env]
----
# ...
transaction.state.log.replication.factor=3
transaction.state.log.min.isr=2
# ...
----

Similarly, the internal  `__consumer_offsets` topic, which stores partition offset commits, has default settings for the number of partitions, replication factor, and acknowledgments.

> Acknowledgments defaults to -1, which means that messages are not acknowledged until replicas for the topic are in sync.

Author: @tombentley  Shouldn't we say what -1 means? _Acknowledgments (_acks_) defaults to -1, which should not be changed.  The -1 setting means that all acknowledgments are required before a commit is accepted._

> Acknowledgments defaults to -1, which means that messages are not acknowledged until replicas for the topic are in sync.

Author: I've removed this line as we're removing the config

[source,env]
----
# ...
offsets.topic.num.partitions=50
offsets.topic.replication.factor=3

> offsets.commit.required.acks=-1

Author: I've also removed the text around this related to acks

# ...
----
<1> The number of network threads for the Kafka cluster.
<2> The number of requests allowed in the request queue.
<3> The number of I/O  threads for a Kafka broker.
<4> The number of threads used for log loading at startup and flushing at shutdown.

NOTE: Kafka broker metrics can help with working out the number of threads required.
For example, metrics for the average time network threads are idle (`kafka.network:type=SocketServer,name=NetworkProcessorAvgIdlePercent`) indicate the percentage of resources used.
If there is 0% idle time, all resources are in use, which means that more threads would be beneficial.

If threads are slow or limited due to the number of disks, you can try increasing the size of the buffers for network requests to improve throughput:

[source,env]
----
# ...
replica.socket.receive.buffer.bytes=65536
# ...
----

And also increase the maximum number of bytes Kafka can receive:

[source,env]
----
# ...
socket.request.max.bytes=104857600
# ...
----

== Increasing bandwidth for high latency connections

Kafka batches data to achieve reasonable throughput over high-latency connections from Kafka to clients, such as connections between datacenters.

> However, if you are experiencing levels of latency that are impairing performance unduly, you can increase the size of the buffers for sending and receiving messages.

Author: Severely!

[source,env]
----
# ...
log.cleaner.dedupe.buffer.size=134217728
# ...
----

Each log entry uses exactly 24 bytes, so you can work out how many log entries the buffer can handle in a single run and adjust the setting accordingly.

If possible, consider increasing the number of log cleaner threads if you are looking to reduce the log cleaning time:

[source,env]
----
# ...
log.cleaner.threads=8
# ...
----

If you are experiencing issues with 100% disk bandwidth usage, you can throttle the log cleaner I/O so that the sum of the read/write operations is less than a specified double value based on the capabilities of the disks performing the operations:

[source,env]
----
# ...
log.cleaner.io.max.bytes.per.second= 1.7976931348623157E308
# ...
----

== Handling large message sizes

The default batch size for messages is 1MB, which is optimal for maximum throughput in most use cases.
Kafka can handle larger batches at a reduced throughput, assuming adequate disk capacity.

Larger messages are handled in four ways:

. Reference-based messaging sends only a reference to data stored in some other system in the message’s value.

> . xref:optimizing_throughput_and_latency[Producer-side message compression] writes compressed messages to the log.

Author: Rearranged

. Inline messaging splits messages into chunks that use the same key, which are then combined on output using a stream-processor like Kafka Streams.
. Broker and producer/consumer client application configuration built to handle larger message sizes.

The reference-based messaging and message compression options are recommended and cover most situations.
With any of theses options, care must be take to avoid introducing performance issues.

.Reference-based messaging

Reference-based messaging is useful for data replication when you do not know how big a message will be.
The external data store must be fast, durable, and highly available for this configuration to work.
Data is written to the data store and a reference to the data is returned.
The producer sends the reference to Kafka.
The consumer uses the reference to fetch the data from the data store.

.Reference-based messaging flow
image::tuning/broker-tuning-messaging-reference.png[Image of reference-based messaging flow]

> As the message passing requires more trips, end-to-end latency will increase.

Author: Nice addition

A hybrid approach would be to only send large messages to the data store and process standard-sized messages directly.

.Producer-side compression

For producer configuration, you specify a `compression.type`, such as Gzip, which is then applied to batches of data generated by the producer.
Using the broker configuration `compression.type=producer`, the broker retains the compression.
Compression adds additional processing overhead on the producer and decompression overhead on the consumer, but includes more data in a batch.

Combine producer-side compression with fine-tuning of the batch size to facilitate optimum throughput.
Using metrics helps to gauge the average batch size needed.

.Inline messaging

Inline messaging is complex, but it does does not have the overhead of depending on external systems like reference-based messaging.

The client application has to serialize and then chunk the data if the message is too big.
The producer then uses the Kafka `ByteArraySerializer` or similar to serialize each chunk again before sending it.

> The consumer receives the chunks, which are assembled before deserialization. Offsets are made for each chunk.

Author: That each chunk has a separate offset position. I've removed it.

The consumer tracks messages and buffers chunks until it has a complete message.
Complete messages are delivered in order according to the offset of the first or last chunk for each set of chunked messages.
Successful delivery of the complete message is checked against offset metadata to avoid duplicates during a rebalance.

.Inline messaging flow
image::tuning/broker-tuning-messaging-inline.png[Image of inline messaging flow]

Inline messaging has a performance overhead on the consumer side because of the buffering required, particularly when handling a series of large messages in parallel.
The chunks of large messages can become interleaved, so that it  is not always possible to commit when all the chunks of a message have been consumed if the chunks of another large message in the buffer are incomplete.
For this reason, the buffering is usually supported by persisting message chunks or by implementing commit logic.

.Configuration to handle larger messages

If larger messages cannot be avoided, and to avoid blocks at any point of the message flow, you can increase the message limits with broker configuration.

> Use `message.max.bytes` to set the maximum record batch size (which can also be set in topic configuration using `max.message.bytes`).

Author: Changed to:

_To do this, configure `message.max.bytes` at the topic level to set the maximum record batch size for individual topics.
If you set `message.max.bytes` at the broker level, larger messages are allowed for all topics. _

The broker will reject any message that is greater than the limit set with `message.max.bytes`.
The buffer size for the producers (`max.request.size`) and consumers (`message.max.bytes`) must be able to accommodate the larger messages.

== Controlling the log flush of message data

Log flush properties control the periodic writes of cached message data to disk.
The scheduler specifies the frequency of checks on the log cache in milliseconds:

[source,env]
----
# ...
log.flush.scheduler.interval.ms=2000
# ...
----

You can control the frequency of the flush based on the maximum amount of time that a message is kept in-memory and the maximum number of messages in the log before writing to disk:

[source,env]
----
# ...
log.flush.interval.ms=50000
log.flush.interval.messages=100000
# ...
----

> To improve latency, you can increase the frequency of log flushes, so that message data is made available to consumers more quickly.

Author: I've removed this sentence

You can adjust the lag time before a follower is considered out of sync:

[source,env]
----
# ...
replica.lag.time.max.ms
# ...
----

This is the latency you will accept between the leader and followers that are considered in sync. You can decrease the time to reduce the lag, but by doing so you might increase the number of followers that fall out of sync. The right value depends on both network latency and broker disk bandwidth.

NOTE: With a high lag time data might be lost in certain circumstances.
Suppose your partition leader fails and followers have not caught up with the latest messages.

> Messages not flushed in the period between a follower sending its next request for messages (while it is still considered in sync) and transition to a new leader might be lost.

Author: I've removed the note considering the doubts.
For the lag time config, I've distilled to this:

_Lag time puts an upper limit on the time to replicate a message to all in-sync replicas and how long a follower can remain in sync. If a follower fails to make a fetch request and catch up with the latest message within the specified lag time, it is removed from in-sync replicas. You can reduce the lag time to detect failed replicas sooner, but by doing so you might increase the number of followers that fall out of sync needlessly. The right lag time value depends on both network latency and broker disk bandwidth._

The first broker in a partition’s list of replicas is known as the preferred replica.  Kafka tries to ensure that, on average, each broker is the _preferred_ replica for a similar number of partitions.   If a leader fails, this affects the balance of a Kafka cluster (as does the assignment of partition replicas to brokers).

By default, Kafka is enabled for automatic partition leader rebalancing based on a periodic check of leader distribution. That is, Kafka checks to see if the preferred leader is the current leader.
A rebalance ensures that leaders are evenly distributed across brokers and brokers are not overloaded.
You can control the frequency, in seconds, of the rebalance check and the maximum percentage of imbalance allowed for a broker before a rebalance is triggered.

[source,env]
----
#...
auto.leader.rebalance.enable=true
leader.imbalance.check.interval.seconds=300
leader.imbalance.per.broker.percentage=10
#...
----

The percentage imbalance for a broker is the gap between the current number of partition leaders it holds and the number of partitions which are preferred leaders. You can set the percentage to zero to ensure that preferred leaders are always elected.

> If the checks for automated rebalances need more control, you can disable it. You can then choose to trigger a rebalance.

Author: Replaced by `kafka-leader-election.sh`?

NOTE: The Grafana dashboards provided with Strimzi show metrics for under-replicated partitions and partitions that do not have an active leader.

[id='con-broker-config-properties-unclean-{context}']
== Unclean leader election

Leader election to an in-sync replica is considered clean because it guarantees no loss of data. But what if there is no in-sync replica to take on leadership?
If a minimum number of in-sync replicas is not set, and there are no followers in sync with the partition leader when its hard drive fails irrevocably, data is already lost.
Not only that, but a new leader cannot be elected because there are no in-sync followers.

You can configure how Kafka handles leader failure:

[source,env]
----
# ...
unclean.leader.election.enable=false
# ...
----

Unclean leader election is disabled by default, which means that out-of-sync replicas cannot become leaders.
Kafka waits until the original leader is back online before messages are picked up again.
Unclean leader election means out-of-sync replicas can become leaders, but you risk losing messages.

> The choice you make depends on whether your requirements favor availability or durability. If you cannot afford the risk of data loss, then leave the default configuration.

Reviewer: We should mention that this can be overridden at the topic level via https://kafka.apache.org/documentation.html#topicconfigs_unclean.leader.election.enable

== Avoiding unnecessary consumer group rebalances

> If new consumers are joining a consumer group that is making requests on a broker, you can add a delay so that unnecessary rebalances to the broker are avoided:

Reviewer: I don't think we really need the "that is making requests on a broker" part. And this isn't really _adding_ a delay, rather the delay is an intrinsic part of how the coordinator works. 

I think something like this is all that needs to be said:

When a group is being created, or is rebalancing, this is the amount of time that the coordinator will wait for members to join the group. The longer this is, the more likely it is that all the members will join in time (thus avoiding a rebalance), but the longer the group will be prevented from actually consuming.

Author: Changed to:

_When creating or rebalancing a group, the delay is the amount of time that the coordinator waits for members to join. The longer the delay, the more likely it is that all the members will join in time and avoid a rebalance.
But the delay also prevents the group from consuming until the period has ended._

// This module is included in the following files:
//
// assembly-client-config.adoc

[id='con-broker-config-properties-{context}']
= Kafka broker configuration tuning

[role="_abstract"]
Use configuration properties to optimize the performance of Kafka brokers.
You can use standard Kafka broker configuration options, except for properties managed directly by Strimzi.

== Basic broker configuration
Certain broker configuration options are managed directly by Strimzi, driven by the `Kafka` custom resource specification:

* `broker.id` is the ID of the Kafka broker
* `log.dirs` are the directories for log data
* `zookeeper.connect` is the configuration to connect Kafka with ZooKeeper

> * `listener` exposes the Kafka cluster to external clients

Reviewer 2: Maybe "external" could be misleading here? External to what? Kubernetes? Clients can even run on the same Kubernetes cluster and connecting through listeners. I would just avoid to use "external".

# ...
----

You can estimate the optimal size of your buffers using a _bandwidth-delay product_ calculation,
which multiplies the maximum bandwidth of the link (in bytes/s) with the round-trip delay (in seconds) to give an estimate of how large a buffer is required to sustain maximum throughput.

== Managing logs with data retention policies

Kafka uses logs to store message data. Logs are a series of segments.
Data is written to an _active_ segment, which receives new messages and is never deleted.
Periodically, a new active segment is _rolled_.
Older segments are retained until they are eligible for deletion.

You can set the maximum size of a log segment:

[source,env]
----
# ...

> log.segment.bytes=1073741824

Author: Okay. So now we have:

> Default settings at the broker level set the maximum size in bytes of a log segment and the amount of time in milliseconds before an active segment is rolled:
> 
> ```
> # ...
> log.segment.bytes=1073741824
> segment.ms=604800000
> # ...
> ```
> 
> You can override these settings at the topic level. 
> Whether you need to lower or raise these values depends on the policy for segment deletion.

> log.segment.bytes=1073741824

Reviewer: Not quite. The situation is a bit asymmetric:

<table>
<tr><th>Broker config</th><th>Topic config</th></tr>
<tr><td>log.roll.ms</td><td>segment.ms</td></tr>
<tr><td>log.segment.bytes</td><td>segment.bytes</td></tr>
</table>

So we need to change the example to use `log.roll.ms` and 

"You can override these settings at the topic level using `segment.bytes` and `segment.ms`."

> log.segment.bytes=1073741824

Author: Ah, that's the topic config one. 
Keeping the setting at 7 days, changed to:

> Default settings at the broker level set the maximum size in bytes of a log segment and the amount of time in milliseconds before an active segment is rolled:
> 
> ```
> # ...
> log.segment.bytes=1073741824
> log.roll.ms=604800000
> # ...
> ```
> 
> You can override these settings at the topic level using `segment.bytes` and `segment.ms`.

In which case, you can set policy to compact _and_ delete logs.
If you choose to compact and delete, first the log data is compacted, removing records with a key in the head of the log.
After which, data that falls before the log retention threshold is deleted.

.Log retention point and compaction point
image::tuning/broker-tuning-compaction-retention.png[Image of compaction with retention point]

You set the frequency the log is checked for cleanup in milliseconds:

[source,env]
----
# ...
log.retention.check.interval.ms=300000
# ...
----

Adjust the log retention check interval in relation to the log retention settings.
Smaller retention sizes might require more frequent checks.

The frequency of cleanup should be often enough to manage the disk space, but not so often it affects performance on a topic.

You can also set a time in milliseconds to put the cleaner on standby if there are no logs to clean:

[source,env]
----
# ...
log.cleaner.backoff.ms=15000
# ...
----

If you choose to delete older log data, you can set a period in milliseconds to retain the deleted data before it is purged:

[source,env]
----
# ...
log.cleaner.delete.retention.ms=86400000
# ...
----

> 

Author: worth mentioning :+1:

== Handling large message sizes

> The default batch size for messages is 1MB, which is optimal for maximum throughput in most use cases.

Reviewer 3: LGTM :)

# ...
----

The wait between flushes includes the time to make the check and the specified interval before the flush is carried out.
Increasing the frequency of flushes can affect throughput.

Generally, the recommendation is to not set explicit flush thresholds and let the operating system perform background flush using its default settings.
Partition replication provides greater data durability than writes to any single disk as a failed broker can recover from its in-sync replicas.

If you are using application flush management, setting lower flush thresholds might be appropriate if you are looking at ways to decrease latency or you are using faster disks.

== Partition rebalancing for availability

Partitions can be replicated across brokers for fault tolerance.
For a given partition, one broker is elected leader and handles all producer requests (writes to the log).
Partition followers on other brokers replicate the partition data of the partition leader for data reliability in the event of the leader failing.

> Followers do not normally serve clients, though xref:type-Rack-reference[`rack` configuration] allows a consumer to consume messages from the closest replica when a Kafka cluster spans multiple datacenters.

Author: The link has a section on _Consuming messages from the closest replicas_. This includes `RackAwareReplicaSelector` and the selector plugin that enables clients to consume from the nearest replica

https://strimzi.io/docs/operators/in-development/using.html#consuming_messages_from_the_closest_replicas

> Followers do not normally serve clients, though xref:type-Rack-reference[`rack` configuration] allows a consumer to consume messages from the closest replica when a Kafka cluster spans multiple datacenters.

Reviewer: Perfect!

Followers operate only to replicate messages from the partition leader and allow recovery should the leader fail.
Recovery requires an in-sync follower. Followers stay in sync by sending fetch requests to the leader, which returns messages to the follower in order.
The follower is considered to be in sync if it has caught up with the most recently committed message on the leader.
The leader checks this by looking at the last offset requested by the follower.
An out-of-sync follower is usually not eligible as a leader should the current leader fail, unless xref:con-broker-config-properties-unclean-{context}[unclean leader election is allowed].

You can adjust the lag time before a follower is considered out of sync:

[source,env]
----
# ...
replica.lag.time.max.ms=30000
# ...
----

> Lag time puts an upper limit on the time to replicate a message to all in-sync replicas and how long a follower can remain in sync.

Reviewer: ```suggestion
Lag time puts an upper limit on the time to replicate a message to all in-sync replicas and how long a producer has to wait for an acknowledgement.
```

If a follower fails to make a fetch request and catch up with the latest message within the specified lag time, it is removed from in-sync replicas.
You can reduce the lag time to detect failed replicas sooner, but by doing so you might increase the number of followers that fall out of sync needlessly.
The right lag time value depends on both network latency and broker disk bandwidth.

When a leader partition is no longer available, one of the in-sync replicas is chosen as the new leader.

The first broker in a partition’s list of replicas is known as the preferred replica.

> Kafka tries to ensure that, on average, each broker is the _preferred_ replica for a similar number of partitions.

Reviewer: Let's drop this, and it's a bit more complex than this. We can use CC to ensure that, _when_ every partition is lead by its preferred replica, the cluster is approximately optimally balanced. CC does this by figuring out which broker is the best to be the preferred leader for each partition, and reassigning partitions/changing the preferred leader to achieve this. Kafka doesn't do this on its own (all it does try to assign new partitions "evenly", which is not the same thing because different partitions can impart very different loads on brokers). 

_If_ the preferred leaders are optimally arranged then the Kafka feature of ensuring that the current leader is the preferred leader (as much as possible) is useful. If they preferred leader are badly arranged then the Kafka feature achieves nothing.

```suggestion
```

# ...
----

When creating or rebalancing a group, the delay is the amount of time that the coordinator waits for members to join. The longer the delay,
the more likely it is that all the members will join in time and avoid a rebalance.

> But the delay also prevents the group from consuming until the period has ended.

Reviewer: It's not quite this simple. But it's probably best to do that elsewhere than in the broker configs doc.
